### PR TITLE
fix(vanilla): should mount once with atom creator atom

### LINKED ISSUE #2314: mount is executed twice
## Summary
I'm using jotai/vanilla.
In my code the mount is executed twice and that is leading to the bug.

## Code to reproduction
```typescript
const a = atom<number>(0);
a.onMount = function (setAtom) {
    console.log('mount');
    setAtom(1);
    return function () {
        console.log('unmount');
    };
};
const b = atom<void>(function (get) {
    get(atom<number>((get) => get(a)));
});

const store = createStore();
store.sub(b, function () {
    store.get(b);
});
```

### PR DESCRIPTION
**Pull request by nogaten**
## Related Issues or Discussions

Fixes #2314

## Summary
I'm sorry about the previous PR.
Changed evaluation of canUnmountAtom in mountDependencies function after mountAtom.

## Check List

- [ ] `yarn run prettier` for formatting code and docs

### CODE REVIEW

**Review by himself65 — COMMENTED** [review]
Can you add the test case?

**Review by dai-shi — CHANGES_REQUESTED** [review]
Please add the test case: https://github.com/pmndrs/jotai/issues/2314#issuecomment-1873058709

**Comment by nogaten** [comment]
Of course @himself65, @dai-shi. Added test.

**Review by himself65 — COMMENTED** [review]
lgtm

**Review by himself65 — APPROVED** [review]

**Comment by himself65** [comment]
some type error might need to fix

File: tests/vanilla/store.test.tsx [hunk 1]
```diff
@@ -488,3 +488,17 @@ it('should not recompute a derived atom value if unchanged (#2168)', async () =>
   expect(store.get(derived2Atom)).toBe(0)
   expect(derive2Fn).toHaveBeenCalledTimes(1)
 })
+
+it('should mount once with atom creator atom (#2314)', async () => {
+  const countAtom = atom(1)
+  countAtom.onMount = vi.fn((setAtom) => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:494]: Please explicitly type `setAtom` for older TS tests.

**Review by dai-shi — CHANGES_REQUESTED** [review]
@nogaten can you fix CI errors?

File: tests/vanilla/store.test.tsx (hunk 1 above)
Comment (dai-shi) [reply tests/vanilla/store.test.tsx:494]:
```suggestion
  countAtom.onMount = vi.fn((setAtom: (v: number) => void) => {
```

**Review by dai-shi — APPROVED** [review]

**Pull request merged**